### ISSUE TITLE
Watching for new questions with narrow language tags that are missing the generic language tag

### ISSUE THREAD

**Issue by adeak** [SOBotics/Nursery#10]
**Name of the bot:** -
**General topic:** tags
**Requires feedback:** no
**Problem the bot is supposed to solve:**  There are a few tags that have specific versions, but should always imply tagging with a broader tag too. The community has been asking on meta for hierarchical tags; it's not likely to happen in 6-8. So now when a question is only tagged with a specific tag (say, [python-3.x]) but not the generic tag (say, [python]), gold badge holders of the generic tag won't be able to dupe hammer the question, and adding the tag themselves would void their hammer on the question. It would be nice to be able to step in when a question is new and pop the generic tag onto it.

**Ideas for implementation:** the listening part seems straightforward enough: watch the front page, look up tags, post a message if a narrow tag is missing the generic. The hard part is defining tag hierarchies that are likely (preferably guaranteed) to make sense. Combination of tags that _sometimes_ (but not always) go hand-in-hand might cause more harm than good.

**Pros**: the feature doesn't really need a separate bot; it could be included in any existing bots that monitor the front page. There's very little functionality to implement, we only need to get the tags right.

**Cons**: due to my personal experience I'm only aware of [python-2.7],[python-3.x] < [python]. There's [perl5] and [perl6] which _look_ like it is a similar case, but the two language versions are different enough that I suspect the perl guys don't want to slap a [perl] tag on anything with either of the aforementioned tags. While other (not necessarily language-related) examples are sure to be around, it's entirely possible that there aren't enough cases for strict hierarchical tags that would merit creating a bot in the first place.

**Comment by mvarendorff** [comment]
[sql] is another case where there is a bunch of "subtags". Generally I do agree though that there might not be enough tags where this might be useful.

**Comment by Bhargav-Rao** [comment]
We first need to come up with a list of tags which need a main tag when a subtag is present. 

Also @NobodyNada ... given that FireAlarm reads all questions on creation, can this be added as a new filter to Fire?

**Comment by NobodyNada** [comment]
@Bhargav-Rao Absolutely.  It shouldn’t be difficult; I just need to find the time to work in it.

**Comment by Filnor** [comment]
Since there was no update to this for quite some while and after getting the idea from @Bhargav-Rao, I dug into Java and wrote [a bot](https://github.com/SOBotics/TagGuard) that has this functionality in a very basic way. Currently, the messages posted look like [this](https://chat.stackoverflow.com/rooms/163468/conversation/tagguard-demo). The [tag list](https://github.com/SOBotics/TagGuard/blob/master/data/tags.csv) is just a random list currently, and needs some adjustment so that the "parent" tag is the first entry and everything that follows are the "child" tags.

**Comment by ztane** [comment]
C, C++ standard versions require a tag too, but they're not as common alone. But also there are lots of common frameworks and libraries that exist only for a specific language, yet the question lacks the proper language tag.

**Comment by Filnor** [comment]
After a very long time, I finally implemented it as a part of of CheckYerFlags: https://github.com/SOBotics/CheckYerFlags
